cdc/puller/sorter/merger.go

```diff
@@ -417,3 +482,89 @@ func printError(err error) error {
 	}
 	return err
 }
+
+// taskBuffer is used to store pending flushTasks.
+// The design purpose is to reduce the backpressure caused by a congested output chan of the merger,
+// so that heapSorter does not block.
+type taskBuffer struct {
+	mu       sync.Mutex
+	queue    deque.Deque
+	notifier notify.Notifier
+	len      *int64
+	isClosed int32
+}
+
+func newTaskBuffer(len *int64) *taskBuffer {
+	return &taskBuffer{
+		queue:    deque.NewDeque(),
+		notifier: notify.Notifier{},
+		len:      len,
+	}
+}
+
+func (b *taskBuffer) put(task *flushTask) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.queue.PushBack(task)
+	prevCount := atomic.AddInt64(b.len, 1)
+
+	if prevCount == 0 {
+		b.notifier.Notify()
+	}
+}
+
+func (b *taskBuffer) get(ctx context.Context) (*flushTask, error) {
+	if atomic.LoadInt32(&b.isClosed) == 1 && atomic.LoadInt64(b.len) == 0 {
+		return nil, nil
+	}
+
+	if atomic.LoadInt64(b.len) == 0 {
+		recv, err := b.notifier.NewReceiver(time.Millisecond * 50)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		defer recv.Stop()
+
+		startTime := time.Now()
+		for atomic.LoadInt64(b.len) == 0 {
+			select {
+			case <-ctx.Done():
+				return nil, errors.Trace(ctx.Err())
+			case <-recv.C:
+			}
+
+			if atomic.LoadInt32(&b.isClosed) == 1 {
+				return nil, nil
+			}
+
+			if time.Since(startTime) > time.Second*5 {
+				log.Debug("taskBuffer reading blocked for too long", zap.Duration("duration", time.Since(startTime)))
+			}
+		}
+	}
+
+	postCount := atomic.AddInt64(b.len, -1)
```

Comment:
for above `put`, you handle `b.len` and queue in the mutex together, but here you handle `b.len` outside of mutex, is it correct?

Reply:
The mutex is used to protect the queue but not `b.len`, so there is no problem. Moreover, since `atomic` operations in golang's concurrency model enforces a happens-before relation, the enqueuing and bumping the `len` operations have a strictly defined order, so it is okay to assume we see the bumped `len` only after enqueuing has been successful even without a mutex.